src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,235 @@
+        }
+
+        private List<ProjectSystemReferenceInfo> GetReferences()
+        {
+            var references = new List<ProjectSystemReferenceInfo>();
+
+            references.AddRange(_projectAbstractReferenceHandler.GetReferences());
+            references.AddRange(_packageAbstractReferenceHandler.GetReferences());
+            references.AddRange(_assemblyAbstractReferenceHandler.GetReferences());
+
+            return references;
+        }
+
+        public async Task<bool> TryUpdateReferenceAsync(string projectPath, string targetFrameworkMoniker, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+        {
+            bool wasUpdated = false;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.None)
+            {
+                return wasUpdated;
+            }
+
+            var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+            var referenceHandler = _mapReferenceTypeToHandler[referenceUpdate.ReferenceInfo.ReferenceType];
+
+            if (referenceHandler != null)
+            {
+                wasUpdated = await ApplyAction(referenceUpdate, cancellationToken, referenceHandler, activeConfiguredProject);
+            }
+
+            return wasUpdated;
+        }
+
+        private static async Task<bool> ApplyAction(ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken,
+            AbstractReferenceHandler abstractReferenceHandler, ConfiguredProject activeConfiguredProject)
+        {
+            bool wasUpdated = false;
+
+            if (referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ||
+                referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUnused)
+            {
+                wasUpdated =
+                    await abstractReferenceHandler.UpdateReferenceAsync(activeConfiguredProject, referenceUpdate, cancellationToken);
+            }
+            else
+            {
+                if (CanRemoveReference(referenceUpdate, abstractReferenceHandler))
+                {
+                    await abstractReferenceHandler.RemoveReferenceAsync(activeConfiguredProject, referenceUpdate.ReferenceInfo);
+                    wasUpdated = true;
+                }
+            }
+
+            return wasUpdated;
+        }
+
+        private static bool CanRemoveReference(ProjectSystemReferenceUpdate referenceUpdate, AbstractReferenceHandler abstractReferenceHandler)
+        {
+            ProjectSystemReferenceInfo referenceInfo;
+
+            var references = abstractReferenceHandler.GetReferences();
+            try
+            {
+                referenceInfo = references.First(c =>
+                    c.ItemSpecification == referenceUpdate.ReferenceInfo.ItemSpecification);
+            }
+            catch
+            {
+                return false;
+            }
```

Comment:
Avoid using exceptions for control flow -- they have a performance impact and make debugging harder. Could use `FirstOrDefault` here then check for null.

Reply:
done. Try/catch removed